### PR TITLE
FQDN Cache and DNSPoller

### PR DESCRIPTION
**Pull request by eloycoto**
Read per commit: 

1) Usage of min TTL on dnsCache, so it's handle in one place. 
2) Use a dnsHistory in poller, so can be restored in the DNS cache GC operations. 

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7629)

### CODE REVIEW

**Comment by eloycoto** [comment]
test-me-please

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/cache.go [hunk 1]
```diff
@@ -170,15 +168,18 @@ func NewDNSCacheWithLimit(limit int) *DNSCache {
 // ips may be an IPv4 or IPv6 IP. Duplicates will be removed.
 // ttl is the DNS TTL for ips and is a seconds value.
 func (c *DNSCache) Update(lookupTime time.Time, name string, ips []net.IP, ttl int) bool {
+	c.Lock()
+	if c.minTTL > 0 && ttl < c.minTTL {
```
Comment (raybejjani) [inline pkg/fqdn/cache.go]: Since minTTL is a constant, do we need to take the lock? In any case, if you move the lock here could you also move the defer unlock?

File: pkg/fqdn/rulegen.go [hunk 2]
```diff
@@ -85,10 +85,6 @@ type RuleGen struct {
 // When config.Cache is nil, the global fqdn.DefaultDNSCache is used.
 func NewRuleGen(config Config) *RuleGen {
 
-	if config.Cache == nil {
```
Comment (raybejjani) [inline pkg/fqdn/rulegen.go]: This means we would panic if we con't pass `.Cache` in. We have to either return an error when it is missing or make it a required parameter, perhaps by making it a parameter to `NewRuleGen`.

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/rulegen.go (hunk 2 above)
Comment (eloycoto) [reply pkg/fqdn/rulegen.go]: @raybejjani fixed.

File: pkg/fqdn/cache.go (hunk 1 above)
Comment (eloycoto) [reply pkg/fqdn/cache.go]: @raybejjani fixed.

**Comment by ianvernon** [comment]
@eloycoto the description of this PR still says 'WIP cleanup' but is labeled as 'pending-review' - can you update the description please?

**Comment by eloycoto** [comment]
@ianvernon done.

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/cache.go [hunk 3]
```diff
@@ -170,6 +168,10 @@ func NewDNSCacheWithLimit(limit int) *DNSCache {
 // ips may be an IPv4 or IPv6 IP. Duplicates will be removed.
 // ttl is the DNS TTL for ips and is a seconds value.
 func (c *DNSCache) Update(lookupTime time.Time, name string, ips []net.IP, ttl int) bool {
+	if c.minTTL > 0 && ttl < c.minTTL {
```
Comment (raybejjani) [inline pkg/fqdn/cache.go]: This isn't wrong but the operation is effectively `max(c.minTTL, ttl)`. We can make a helper function or skip the `if c.minTTL >0` part.

File: pkg/fqdn/cache.go [hunk 4]
```diff
@@ -137,25 +132,28 @@ type DNSCache struct {
 
 	// perHostLimit is the number of maximum number of IP per host.
 	perHostLimit int
+
+	minTTL int
```
Comment (raybejjani) [inline pkg/fqdn/cache.go:139]: Could you add a comment explaining what this variable is for, and that it is a constant and not protected by the lock. Also explain what 0 means and that the unit is seconds to match the TTL format in DNS. `perHostLimit` is also a constructor-time constant. It might be good to move them above the `lock.RWMutex` line since we have that convention elsewhere (things below the lock are protected, and those above are not). In either case, it might be good to mention that perHostLimit is a constant too, and the meaning of the 0 value.

File: pkg/fqdn/cache_test.go [hunk 5]
```diff
@@ -487,8 +487,39 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
...
+func (ds *DNSCacheTestSuite) TestTTLInsertWithMinValue(c *C) {
+
+	cache := NewDNSCache(60)
+	cache.Update(now, "test.com", []net.IP{net.ParseIP("1.2.3.4")}, 3)
+
+	// Checking just now to validate that is inserted correctly
+	res := cache.lookupByTime(time.Now(), "test.com")
+	c.Assert(res, HasLen, 1)
+	c.Assert(res[0].String(), Equals, "1.2.3.4")
+
```
Comment (raybejjani) [inline pkg/fqdn/cache_test.go:504]: We need to test at `future=now.Add(time.Second * 3)` (or `*4`) to verify that the minTTL was used and not the original 3s TTL.

File: pkg/fqdn/cache_test.go [hunk 6]
```diff
@@ -487,8 +487,39 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
 	}
 }
 
+func (ds *DNSCacheTestSuite) TestTTLInsertWithMinValue(c *C) {
+
+	cache := NewDNSCache(60)
+	cache.Update(now, "test.com", []net.IP{net.ParseIP("1.2.3.4")}, 3)
```
Comment (raybejjani) [inline pkg/fqdn/cache_test.go:493]: We should declare now inside this function. This global `now`  is part of the benchmarks. Unfortunately some other tests also use it. We should probably fix those too (but that can be a separate PR, or at least commit).

File: pkg/fqdn/cache_test.go [hunk 7]
```diff
@@ -487,8 +487,39 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
...
+
+	// Checking just now to validate that is inserted correctly
+	res := cache.lookupByTime(time.Now(), "test.com")
```
Comment (raybejjani) [inline pkg/fqdn/cache_test.go]: This should use the same now as the `Update` (so, the global `now` in the current code, or a `now` declared inside this function if you agree with my other comment).

File: pkg/fqdn/cache_test.go [hunk 8]
```diff
@@ -487,8 +487,39 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
...
+
+	// Validate that in future time the value is correct
+	future := time.Now().Add(time.Second * 70)
+	res = cache.lookupByTime(future, "test.com")
+	c.Assert(res, HasLen, 0)
+}
+
+func (ds *DNSCacheTestSuite) TestTTLInsertWithZeroValue(c *C) {
+	cache := NewDNSCache(0)
+	cache.Update(now, "test.com", []net.IP{net.ParseIP("1.2.3.4")}, 10)
```
Comment (raybejjani) [inline pkg/fqdn/cache_test.go:514]: The now usage should be changed like the comments for `TestTTLInsertWithMinValue`

File: pkg/fqdn/cache_test.go [hunk 9]
```diff
@@ -487,8 +487,39 @@ func benchmarkUnmarshalJSON(c *C, numDNSEntries int) {
...
+
+	// Checking just now to validate that is inserted correctly
+	res := cache.lookupByTime(time.Now(), "test.com")
+	c.Assert(res, HasLen, 1)
+	c.Assert(res[0].String(), Equals, "1.2.3.4")
+
```
Comment (raybejjani) [inline pkg/fqdn/cache_test.go:525]: We need to test with now.Add(time.Second) as well, in case the implementation uses the minTTL of 0 and expires it in the next second. This is somewhat covered by the check above, though, so it isn't super important.

File: pkg/fqdn/cache.go [hunk 10]
```diff
@@ -137,25 +132,28 @@ type DNSCache struct {
...
 }
 
 // NewDNSCache returns an initialized DNSCache
-func NewDNSCache() *DNSCache {
+func NewDNSCache(minTTL int) *DNSCache {
```
Comment (raybejjani) [inline pkg/fqdn/cache.go:143]: It might be worthwhile for us to create a constant `NoMinTTL` or something. We originally argued against `NoPerHostIPLimit` because `NewDNSCacheWithLimit` was more clear, but now we are mixing philosophies a bit (admittedly, minTTL is required now, while limit is optional).

File: pkg/fqdn/dnspoller_test.go [hunk 11]
```diff
@@ -158,7 +158,7 @@ func (ds *FQDNTestSuite) TestRuleGenRuleHandling(c *C) {
 			}
 
 			gen    = NewRuleGen(cfg)
-			poller = NewDNSPoller(cfg, gen)
+			poller = NewDNSPoller(cfg, gen, NewDNSCache(0))
```
Comment (raybejjani) [inline pkg/fqdn/dnspoller_test.go]: This doesn't set a limit on the poller, which was the point, no? It makes a new empty cache with no limit.

File: pkg/fqdn/dnspoller.go [hunk 12]
```diff
@@ -60,10 +60,14 @@ type DNSPoller struct {
...
 	config Config
+
+	// DNSHistory is the collection of still-valid DNS responses intercepted
+	// for the poller
+	DNSHistory *DNSCache
 }
 
 // NewDNSPoller creates an initialized DNSPoller. It does not start the controller (use .Start)
-func NewDNSPoller(config Config, ruleManager *RuleGen) *DNSPoller {
+func NewDNSPoller(config Config, ruleManager *RuleGen, DNSHistory *DNSCache) *DNSPoller {
```
Comment (raybejjani) [inline pkg/fqdn/dnspoller.go]: Whey do we need to pass `DNSHistory` in? It's internal to the poller (it's only needed to enforce the limit and to update the global cache during TTL expiration or forced expiration.

File: daemon/fqdn.go [hunk 13]
```diff
@@ -536,7 +533,7 @@ func extractDNSLookups(endpoints []*endpoint.Endpoint, CIDRStr, matchPatternStr
 	return lookups, nil
 }
 
-func deleteDNSLookups(endpoints []*endpoint.Endpoint, expireLookupsBefore time.Time, matchPatternStr string) (namesToRegen []string, err error) {
+func deleteDNSLookups(cache *fqdn.DNSCache, endpoints []*endpoint.Endpoint, expireLookupsBefore time.Time, matchPatternStr string) (namesToRegen []string, err error) {
```
Comment (raybejjani) [inline daemon/fqdn.go]: it might be good to call this `globalCache` or something more clear. It's probably worth it to explain all the parameters in a comment too.

File: pkg/fqdn/dnspoller.go [hunk 14]
```diff
@@ -76,9 +80,14 @@ func NewDNSPoller(config Config, ruleManager *RuleGen) *DNSPoller {
 		config.PollerResponseNotify = noopPollerResponseNotify
 	}
 
+	if DNSHistory == nil {
```
Comment (raybejjani) [inline pkg/fqdn/dnspoller.go]: We can probably stop using the config.Cache variable here. We still call `Rulegen.Update`, and we would instead do that in `PollerNotifyResponse` someday. Not using it would be consistent with what we ended up with (despite the comment on `Config.Cache`, which we should update).

File: daemon/fqdn.go [hunk 15]
```diff
@@ -133,7 +132,10 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
 		}}
 
 	d.dnsRuleGen = fqdn.NewRuleGen(cfg)
-	d.dnsPoller = fqdn.NewDNSPoller(cfg, d.dnsRuleGen)
+	pollerDNSCache := fqdn.NewDNSCacheWithLimit(
```
Comment (raybejjani) [inline daemon/fqdn.go]: Was there a reason to do this like this? instead of using poller.GetDNSCache() ? I guess you wanted to avoid passing variables to `NewDNSPoller`? We pass cfg in, and the perHostIPs limit and the minTTL should live there, so that's one place to pass them through. In fact, since we create cfg just above, this is a good place to handle setting the defaults for minTTL and for perHostIPs and simply pass that around.

**Review by raybejjani — CHANGES_REQUESTED** [review]

File: daemon/fqdn.go (hunk 15 above)
Comment (eloycoto) [reply daemon/fqdn.go]: If we use the same poller as the general one, the DNS History will not work, because on GC we deleted, and after that we restore, in this case nothing to restore.

File: pkg/fqdn/dnspoller.go (hunk 12 above)
Comment (eloycoto) [reply pkg/fqdn/dnspoller.go]: Addressed in the comment in `daemon/fqdn.go`

File: daemon/fqdn.go (hunk 15 above)
Comment (raybejjani) [reply daemon/fqdn.go]: Oh, I meant that we create `pollerDNSCache` here to use it later in the code, instead of using `d.poller.GetDNSCache()` when we do the GC. I would expect that the global DNSCache and the DNSCache (poller.DNSHistory) are still separate things.

File: daemon/fqdn.go (hunk 13 above)
Comment (raybejjani) [reply daemon/fqdn.go]: Alternatively, if you meant for this to be the poller cache, we should call it that. Then, we would need to do something different on line 552 since it now uses the wrong cache.

File: pkg/fqdn/dnspoller.go (hunk 12 above)
Comment (raybejjani) [reply pkg/fqdn/dnspoller.go]: I don't think I understand (but I did reply to https://github.com/cilium/cilium/pull/7629/files#r274884582). The poller needs cache specific to it, with a perHostIP limit. This can't be the global cache (which this PR also removes). We can create the global cache in `BootstrapFQDN()` since it now needs to know the minTTL but the poller doesn't need a reference to that.

File: pkg/fqdn/cache.go [hunk 16]
```diff
@@ -26,11 +26,6 @@ import (
 	"github.com/cilium/cilium/pkg/lock"
 )
 
-// DefaultDNSCache is a global, shared, DNS cache. It is the default cache used
-// by DNSPoller instances, unless initialized to use another.
-// Note: The DNSCache type returns all DNS information, regardless of source.
-var DefaultDNSCache = NewDNSCache()
```
Comment (raybejjani) [inline pkg/fqdn/cache.go:32]: This might break some other things. I remember some monitor code referring to this directly. I don't think we're using it but hopefully CI compiles all important things.

File: pkg/fqdn/cache.go [hunk 17]
```diff
@@ -170,6 +168,10 @@ func NewDNSCacheWithLimit(limit int) *DNSCache {
 // ips may be an IPv4 or IPv6 IP. Duplicates will be removed.
 // ttl is the DNS TTL for ips and is a seconds value.
 func (c *DNSCache) Update(lookupTime time.Time, name string, ips []net.IP, ttl int) bool {
+	if c.minTTL > 0 && ttl < c.minTTL {
+		ttl = c.minTTL
```
Comment (raybejjani) [inline pkg/fqdn/cache.go:175]: I just realised that it would be better to also keep a field with the original request TTL. This way, when we restore with a different minTTL we would do the right thing at restore time, instead of reusing the older minTTL which may be too high or inappropriate. This does mean changing updateFromCache to use this new field, instead of the current `.ttl` field. This might also mean that this `if ttl < c.minTTL` check would have to move to `.updateWithEntry`

Comment (eloycoto) [reply pkg/fqdn/cache.go:175]: we spoke offline about this, will do this in a future PR.

File: pkg/fqdn/dnspoller_test.go (hunk 11 above)
Comment (eloycoto) [reply pkg/fqdn/dnspoller_test.go]: is for the test  only ;-)

File: pkg/fqdn/cache.go (hunk 16 above)
Comment (eloycoto) [reply pkg/fqdn/cache.go:32]: We already discussed this offline.

**Comment by eloycoto** [comment]
test-me-please

File: pkg/fqdn/dnspoller.go [hunk 18]
```diff
@@ -60,6 +60,10 @@ type DNSPoller struct {
 	// config is a copy from when this instance was initialized.
 	// It is read-only once set
 	config Config
+
+	// DNSHistory is the collection of still-valid DNS responses intercepted
+	// for the poller
+	DNSHistory *DNSCache
```
Comment (raybejjani) [inline pkg/fqdn/dnspoller.go:67]: Could you also leave a comment here saying this is not protected by the mutex. Perhaps also move it above the mutex since our convention is that everything below is protected.

File: pkg/fqdn/cache.go [hunk 19]
```diff
@@ -137,25 +132,30 @@ type DNSCache struct {
 
 	// perHostLimit is the number of maximum number of IP per host.
 	perHostLimit int
+
+	// minTTL is the minimun TTL value that a cache entry can have, if the TTL
+	// sent in the Update is lower, the TTL will be owerwritten to this value
+	minTTL int
```
Comment (raybejjani) [inline pkg/fqdn/cache.go:139]: Could you also leave a comment here saying minTTL and perHostLImit are not protected by the mutex. Perhaps also move them above the mutex since our convention is that everything below is protected.

**Review by raybejjani — CHANGES_REQUESTED** [review]
Looks good! Just 2 comment changes to avoid confusion.

**Comment by raybejjani** [comment]
Ah, also, because I asked you to change the code and switch `NewDNSCache(Config, *RuleGen, *DNSCache) -> NewDNSCache(Config, *RuleGen)` the unit tests need to also be updated so they compile.

**Comment by eloycoto** [comment]
test-me-please

**Review by ianvernon — APPROVED** [review]

**Review by raybejjani — APPROVED** [review]
